### PR TITLE
Extend query lang

### PR DESCRIPTION
**Pull request by friendlymatthew**
Committing my existing work -- working on e2e testing

### CODE REVIEW

**Comment by friendlymatthew** [comment]
Btw -- sorry about the prettier freaking out. 

My intellij config blew up and I didn't want to spend time fixing that and my VSCode formats it with a lot of fat spaces. 

I'm using Zed and it seems to revert the formatting to intellij conventions

File: src/btree/bptree.ts [hunk 1]
```diff
@@ -3,205 +3,239 @@ import { RangeResolver } from "../resolver";
...
+      throw new Error(`${error}`);
+    }
+  }
+
+  public iter(key: ReferencedValue): TraversalIterator {
+    return new TraversalIterator(this, key);
+  }
+
+  async first(
+  ): Promise<ReferencedValue> {
```
Comment (kevmo314) [inline src/btree/bptree.ts:87]:
```suggestion
  async first(): Promise<ReferencedValue> {
```

File: src/db/database.ts [hunk 2]
```diff
@@ -1,328 +1,297 @@
...
+    if (new Set((query.where ?? []).map((where) => where.key)).size > 1) {
+      throw new Error("composite indexes not supported... yet");
+    }
+
+    const { format } = await this.indexFile.metadata();
+    const dfResolver = this.dataFile.getResolver();
+
+    if (dfResolver === undefined) {
+      throw new Error("data file is undefined");
+    }
```
Comment (kevmo314) [inline src/db/database.ts:71]:
```suggestion
    if (!dfResolver) {
      throw new Error("data file is undefined");
    }
```

File: src/db/database.ts [hunk 3]
```diff
@@ -1,328 +1,297 @@
...
+            if (res === -1) {
+              const [_, mp] = await bptree.find(currentKey);
+
+              const data = await this.dataFile.get(
+                Number(mp.offset),
+                Number(mp.offset) + mp.length - 1
+              );
+              yield handleSelect(data, query.select);
+            } else {
+              console.log("less than")
```
Comment (kevmo314) [inline src/db/database.ts:144]:
```suggestion
```

File: src/db/database.ts [hunk 4]
```diff
@@ -1,328 +1,297 @@
...
+          }
+        } else {
+          const lastKey = await bptree.last();
+          const iter = bptree.iter(lastKey);
+
+          while (await iter.prev()) {
+            const currentKey = iter.getKey();
+
+            let res = ReferencedValue.compareBytes(valueBuf, currentKey.value);
+            if (res === -1 || res === 0) {
```
Comment (kevmo314) [inline src/db/database.ts:176]:
```suggestion
            if (ReferencedValue.compareBytes(valueBuf, currentKey.value) <= 0) {
```

Technically, `compareBytes` could return any negative number, it's not guaranteed to be -1.

File: src/db/database.ts [hunk 5]
```diff
@@ -1,328 +1,297 @@
...
+        }
+      } else if (operation === "<=") {
+        if (ord === "DESC") {
+          const valueRef = new ReferencedValue({ offset: maxUint64, length: 0 }, valueBuf)
+          const iter = bptree.iter(valueRef);
+          while (await iter.prev()) {
+            const currentKey = iter.getKey();
+
+            let res = ReferencedValue.compareBytes(valueBuf, currentKey.value);
+            if (res === 1 || res === 0) {
```
Comment (kevmo314) [inline src/db/database.ts:215]:
```suggestion
            if (ReferencedValue.compareBytes(valueBuf, currentKey.value) >= 0) {
```

File: src/db/database.ts [hunk 6]
```diff
@@ -1,328 +1,297 @@
...
+          }
+        } else {
+          const firstKey = await bptree.first();
+          const iter = bptree.iter(firstKey);
+
+          while (await iter.next()) {
+            const currentKey = iter.getKey();
+
+            let res = ReferencedValue.compareBytes(valueBuf, currentKey.value);
+            if (res === 1 || 0) {
```
Comment (kevmo314) [inline src/db/database.ts:234]:
```suggestion
            if (ReferencedValue.compareBytes(valueBuf, currentKey.value) > 0) {
```

File: src/tests/mock_binaries/bptree_iterator.bin
Comment (kevmo314) [inline src/tests/mock_binaries/bptree_iterator.bin:1]: Remove

File: src/tests/mock_binaries/green_tripdata_2023-01.index
Comment (kevmo314) [inline src/tests/mock_binaries/green_tripdata_2023-01.index:1]: Empty files?

**Review by kevmo314 — APPROVED** [review]

File: src/db/database.ts (hunk 4 above)
Comment (friendlymatthew) [reply src/db/database.ts]: Yeah -- I've updated every single comparison assertion to follow this standard where `<= 0` or `< 0`

For the opposite standard, I've kept it ` >= 0` and `=== 1`

**Pull request merged**